DynamicTablesPkg/Library/Smbios/SmbiosType4Lib/SmbiosType4Generator.c

```diff
@@ -0,0 +1,538 @@
+      SmbiosRecord->Hdr.Type   = EFI_SMBIOS_TYPE_PROCESSOR_INFORMATION;
+      SmbiosRecord->Hdr.Length = sizeof (SMBIOS_TABLE_TYPE4);
+
+      CpuCount    = 0;
+      ThreadCount = 0;
+
+      SmbiosRecord->L1CacheHandle = FindSmbiosHandleEx (
+                                      CREATE_STD_SMBIOS_TABLE_GEN_ID (EStdSmbiosTableIdType07),
+                                      CM_ABSTRACT_TOKEN_MAKE (ETokenNameSpaceSmbios, EStdSmbiosTableIdType07, 0 | (Index << 2))
+                                      );
+      SmbiosRecord->L2CacheHandle = FindSmbiosHandleEx (
+                                      CREATE_STD_SMBIOS_TABLE_GEN_ID (EStdSmbiosTableIdType07),
+                                      CM_ABSTRACT_TOKEN_MAKE (ETokenNameSpaceSmbios, EStdSmbiosTableIdType07, 1 | (Index << 2))
+                                      );
+      SmbiosRecord->L3CacheHandle = FindSmbiosHandleEx (
+                                      CREATE_STD_SMBIOS_TABLE_GEN_ID (EStdSmbiosTableIdType07),
+                                      CM_ABSTRACT_TOKEN_MAKE (ETokenNameSpaceSmbios, EStdSmbiosTableIdType07, 2 | (Index << 2))
+                                      );
+
+      // Find all CPUs on this socket in order to set the cache handles
+      for (CpuIndex = 0; CpuIndex < ProcHierarchyNodeCount; CpuIndex++) {
+        Node  = &ProcHierarchyNodeList[CpuIndex];
+        Flags = (EFI_ACPI_6_3_PPTT_STRUCTURE_PROCESSOR_FLAGS *)&Node->Flags;
+
+        if (Flags->NodeIsALeaf) {
+          // Walk through ParentToken to find the socket node
+          while (!Flags->PhysicalPackage) {
+            Node = FindProcHierarchyInfoFromToken (ProcHierarchyNodeList, ProcHierarchyNodeCount, Node->ParentToken);
+            ASSERT (Node != NULL);
+            Flags = (EFI_ACPI_6_3_PPTT_STRUCTURE_PROCESSOR_FLAGS *)&Node->Flags;
+          }
+
+          // Is this CPU on the current socket?
+          if (Node == &ProcHierarchyNodeList[Index]) {
+            Node  = &ProcHierarchyNodeList[CpuIndex];
+            Flags = (EFI_ACPI_6_3_PPTT_STRUCTURE_PROCESSOR_FLAGS *)&Node->Flags;
+            if (Flags->ProcessorIsAThread) {
+              // Only increment CPU count on the first leaf node
+              for (CpuIndex2 = 0; CpuIndex2 < CpuIndex; CpuIndex2++) {
+                if (ProcHierarchyNodeList[CpuIndex2].ParentToken == Node->ParentToken) {
+                  break;
+                }
+              }
+
+              if (CpuIndex == CpuIndex2) {
+                CpuCount++;
+              }
+            } else {
+              CpuCount++;
+            }
+
+            ThreadCount++;
+          }
+        }
+      }
+
+      SmbiosRecord->ProcessorType = CentralProcessor;
+ #if defined (MDE_CPU_AARCH64)
+      SmbiosRecord->ProcessorFamily          = ProcessorFamilyIndicatorFamily2;
+      SmbiosRecord->ProcessorFamily2         = ProcessorFamilyARMv8;
+      SmbiosRecord->ProcessorId.SocId        = ProcHierarchyNodeList[Index].ProcessorId & 0xffff;

```

Comment:
We could replace the `ProcessorId` type by PROCESSOR_ID_DATA and copy paste the field at this point